### PR TITLE
[Refactor][OutputProcessor 1/3]: clean up multimodal payload accumulation and move payload logic into MultimodalPayload & directory refactor

### PR DESCRIPTION
**Pull request by bowieshi**
PLEASE FILL IN THE PR DESCRIPTION HERE.

### Motivation

`OmniRequestState.add_multimodal_tensor()` and `_consolidate_multimodal_tensors()` in
`vllm_omni/engine/output_processor.py` had grown hard to read: deeply nested if/else branches,
triple-nested `try/except`, repeated type-check patterns, and single-letter variable names
(`k`, `v`, `sk`, `sv`). In addition, all payload normalization/accumulation/consolidation logic
lived in the output processor even though it operates purely on `MultimodalPayload` data.

This PR is a pure refactor — **no behavior change** — that simplifies the two functions and moves
payload-specific logic into the `MultimodalPayload` class where it belongs.

### Changes

**`vllm_omni/engine/mm_outputs.py`**

1. **New classmethod `MultimodalPayload.from_raw(payload, modality_key)`** — normalizes a raw
   producer payload (dict, `MultimodalPayload`, or bare tensor) into a `MultimodalPayload`,
   including CPU transfer and remapping of producer-specific keys (`"model_outputs"`, `"hidden"`)
   to the semantic modality key. Previously inlined in `add_multimodal_tensor()`; the two remap
   branches are collapsed into a single `is_producer_key` condition.
2. **New method `MultimodalPayload.append(incoming)`** — replaces the module-level
   `_merge_payload()` in `output_processor.py`. It owns the empty-payload check and returns the
   resulting payload, so the caller is a single assignment. The two near-identical tensor/metadata
   merge loops are deduplicated into one `_append_entries()` helper.
3. **New methods `consolidate_tensors(strategy)` / `consolidate_metadata()`** — split
   consolidation by semantics: tensors are generated content chunks and are concatenated per
   `TensorAccumulationStrategy`; metadata values are per-step snapshots and resolve to the latest
   value. The supporting helpers (`_is_tensor_list`, `_cat_tensors`, `_consolidate_tensor_list`,
   `_to_cpu`) moved here as module-private functions.

**`vllm_omni/engine/output_processor.py`**

4. **`add_multimodal_tensor()` reduced from ~35 lines to ~15** — now just: resolve modality key →
   `MultimodalPayload.from_raw()` → `MultimodalPayload.append()`.
5. **`_consolidate_multimodal_tensors()` flattened** — the triple-nested `try/except` and
   dict-walking loops are replaced by two method calls (`consolidate_tensors()` +
   `consolidate_metadata()`); the output processor now only resolves the modality and strategy.
6. **Removed dead/duplicated code** — deleted the unused `OmniRequestState._to_cpu` staticmethod
   (no references repo-wide) and the now-unused `TensorAccumulationStrategy` import.
7. **Naming cleanup** — `k/v/sk/sv/t` → `key/value/sub_key/sub_value/chunk`,
   `target_key` → `modality_key`, `_merge_payload` → `append`.

**File directory refactor**
We want to make the file system clear. We moved output related file into same directory called `vllm-omni/outputs`. Related files include `vllm_omni/engine/mm_outputs.py`, `vllm_omni/engine/output_processor.py`, `vllm_omni/outpus.py`, `vllm_omni/engine/output_modality.py`

### Notes

- One intentional micro-tightening while deduplicating the merge loops: appending to an existing
  deferred-concat list now requires the new value to be a tensor (previously the tensors loop
  appended unconditionally). A non-tensor in such a list would have failed at `torch.cat` during
  consolidation anyway, so this only makes the invalid case fail earlier and consistently.

## Test Plan

**vLLM Version: 0.24.0**

**vLLM-Omni Commit: 3a64da194c0a0b3af7448bc6b09aede36cc270eb**

## Test Result
L1 Test: pytest -s -v -m "core_model and cpu" Pass

### CODE REVIEW

File: vllm_omni/engine/mm_outputs.py [hunk 1]
```diff
@@ -11,8 +11,79 @@
...
 import torch
+from vllm.logger import init_logger
 from vllm.outputs import CompletionOutput
 
+from vllm_omni.engine.output_modality import TensorAccumulationStrategy
+
+logger = init_logger(__name__)
+
+
+def _to_cpu(x: Any) -> Any:
```
Comment (tzhouam) [inline vllm_omni/engine/mm_outputs.py]: move to utils

File: vllm_omni/outputs/mm_outputs.py [hunk 2]
```diff
@@ -11,8 +11,79 @@
...
+def _consolidate_tensor_list(
+    key: str,
+    tensor_list: list[torch.Tensor],
+    strategy: TensorAccumulationStrategy,
+) -> torch.Tensor:
+    """Concatenate a deferred tensor list, with fallbacks on shape mismatch."""
+    try:
+        return _cat_tensors(tensor_list, strategy)
+    except RuntimeError:
+        if key != "audio":
```
Comment (tzhouam) [inline vllm_omni/outputs/mm_outputs.py:53]: this part is for async chunk, not for audio only. can we come up with a design that indentify this by whether using async chunk instead of modality?

Comment (bowieshi) [reply vllm_omni/outputs/mm_outputs.py:53]: Can we fix this issue when I deal with _MODALITY_ALIASES issue. Becuase there we will define the final_output_type or some attributes there (like "drainable" or "async_chunk")

File: vllm_omni/engine/mm_outputs.py (hunk 1 above)
Comment (bowieshi) [reply vllm_omni/engine/mm_outputs.py]: already move to `vllm-omni/outputs/utils.py`

File: vllm_omni/outputs/output_processor.py [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+        if "audio" in lowered:
+            return "audio"
+        if "image" in lowered:
+            return "image"
+        if "latent" in lowered:
+            return "latent"
+    return "text"
+
+
+class OmniRequestState(RequestState):
```
Comment (hsliuustc0106) [inline vllm_omni/outputs/output_processor.py:58]: Can we keep the old modules as re-export shims instead of moving by deletion/duplication? `vllm_omni.engine.mm_outputs` and `vllm_omni.engine.output_modality` were documented API paths and now fail to import, while `engine.output_processor` still defines a second copy of these classes, so old and new imports have different class identities.

File: vllm_omni/outputs/output_processor.py [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+
+logger = init_logger(__name__)
+
+
+# ---------------------------------------------------------------------------
+# Module-level helpers
+# ---------------------------------------------------------------------------
+
+
+def _modality_to_type_string(modality: OutputModality) -> str:
```
Comment (hsliuustc0106) [inline vllm_omni/outputs/output_processor.py:36]: does this output modality/processor work for diffusion outputs as well? if so, this should be aligned or unified with diffusion output_formatter/metadata

Comment (hsliuustc0106) [reply vllm_omni/outputs/output_processor.py:36]: cc @bjf-frz

File: vllm_omni/outputs/output_processor.py (hunk 3 above)
Comment (bowieshi) [reply vllm_omni/outputs/output_processor.py:58]: fixed at commit 22ae76941d19b3cf1029374063f8df96b5990e49

File: vllm_omni/outputs/output_processor.py (hunk 4 above)
Comment (bowieshi) [reply vllm_omni/outputs/output_processor.py:36]: I think the type of modality should be made fixed type like enum instead of loss string. This will be refactored in later phase. Some code like following should be replaced by fixed enum and assertion should be put in config part instead of middle in the code:
```python

_MODALITY_ALIASES: dict[str, str] = {
    "speech": "audio",
    "images": "image",
    "latents": "latent",
    "wav": "audio",
    "waveform": "audio",
    "pixel_values": "image",
    "pixels": "image",
    "token_ids": "text",
    "tokens": "text",
}


```

Comment (bowieshi) [reply vllm_omni/outputs/output_processor.py:36]: > does this output modality/processor work for diffusion outputs as well? if so, this should be aligned or unified with diffusion output_formatter/metadata

I have checked #4922, it is compatible with my PR. #4922 modality definition is for diffusionOutput but my definition is for RequestOutput. The intermediate interface is from_diffision function. I think we can merge our two PR into main branch first. So I can continue refactoring phase 2 of outputProcessor (move the alias check in config stage and also coordinate with @bjf-frz for the output_modality definition)

Comment (bowieshi) [reply vllm_omni/outputs/output_processor.py:36]: @hsliuustc0106

**Comment by bowieshi** [comment]
New test results after align with pr #4922 
```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
--- Running Summary
============================================ 4135 passed, 13 skipped, 2174 deselected, 70 warnings in 427.50s (0:07:07) =============================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute

```

**Comment by bowieshi** [comment]
```
====================================================== 4147 passed, 13 skipped, 2189 deselected, 70 warnings in 423.95s (0:07:03) =======================================================
```

File: vllm_omni/engine/mm_outputs.py [hunk 5]
```diff
@@ -1,127 +1,7 @@
...
-        multimodal_output: MultimodalPayload | None = None,
-        **kwargs: Any,
-    ):
-        super().__init__(**kwargs)
-        self.multimodal_output = multimodal_output
-
-    def __repr__(self) -> str:
-        base = super().__repr__()
-        return f"{base[:-1]}, multimodal_output={self.multimodal_output!r})"
+warnings.warn(
```
Comment (lishunyang12) [inline vllm_omni/engine/mm_outputs.py:5]: This shim only warns, it doesn't re-export anything, so `from vllm_omni.engine.mm_outputs import MultimodalPayload` now hard-fails with ImportError instead of warning. Same for the `output_modality`/`output_processor` shims. Add `from vllm_omni.outputs.mm_outputs import *` under the warn so old imports keep working.

File: vllm_omni/outputs/output_processor.py [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+            return
+
+        try:
+            modality = OutputModality.from_string(self.mm_type)
+        except (ValueError, KeyError):
+            modality = OutputModality.TEXT
+        strategy = get_accumulation_strategy(modality)
+
+        try:
+            self.mm_accumulated.consolidate_tensors(strategy)
```
Comment (lishunyang12) [inline vllm_omni/outputs/output_processor.py:128]: Dropping the old `_METADATA_TENSOR_KEYS` relocation is a behavior change — scalar `sr`/`sample_rate`/`audio_sample_rate` tensors used to get moved into `.metadata` before concat. Now they stay in `.tensors`, and if they arrive as multiple chunks they hit the concat fallback plus warning and never land in `.metadata`. Intended? If not, port that relocation into `consolidate_tensors`.

File: vllm_omni/outputs/mm_outputs.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+    def __eq__(self, other: object) -> bool:
+        """Support equality with plain dicts and other Mappings."""
+        if isinstance(other, MultimodalPayload):
+            return self.tensors == other.tensors and self.metadata == other.metadata
+        if isinstance(other, Mapping):
+            return dict(self) == dict(other)
+        return NotImplemented
+
+    def append(self, incoming: MultimodalPayload) -> MultimodalPayload:
```
Comment (lishunyang12) [inline vllm_omni/outputs/mm_outputs.py]: `append` returning `incoming` when empty but `self` otherwise is a footgun — it reads in-place but only works if callers do `x = x.append(y)`. Rename to `merged_with` (or always mutate self) to make that contract obvious.

**Review by lishunyang12 — COMMENTED** [review]
The deprecation shim now ImportErrors instead of warning — that and the dropped tensor->metadata relocation are the real ones.

File: vllm_omni/engine/mm_outputs.py (hunk 5 above)
Comment (bowieshi) [reply vllm_omni/engine/mm_outputs.py:5]: fixed

File: vllm_omni/outputs/output_processor.py (hunk 6 above)
Comment (bowieshi) [reply vllm_omni/outputs/output_processor.py:128]: I currently add the original logic to the the refactor code. I will check the necessary of the `_METADATA_TENSOR_KEYS ` in next refactor phase and determine whether need to remove it.

File: vllm_omni/outputs/mm_outputs.py (hunk 7 above)
Comment (bowieshi) [reply vllm_omni/outputs/mm_outputs.py]: fixed

**Comment by hsliuustc0106** [comment]
I think this is PR is close to merge

**Review by hsliuustc0106 — APPROVED** [review]

File: vllm_omni/outputs/output_modality.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    "speech": "audio",
+    "images": "image",
+    "latents": "latent",
+    "wav": "audio",
+    "waveform": "audio",
+    "pixel_values": "image",
+    "pixels": "image",
+    "token_ids": "text",
+    "tokens": "text",
+}
```
Comment (gcanlin) [inline vllm_omni/outputs/output_modality.py:23]: Is it necessary to support aliases? Seems that it will make more confusion.

Update: oh, it's from the original code. Maybe we can consider clean it in a follow-up PR.

Comment (hsliuustc0106) [reply vllm_omni/outputs/output_modality.py:23]: we can coordinate in follow-up PRs
